akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1402,44 +1402,54 @@ private[stream] object Collect {
 /**
  * INTERNAL API
  */
-@InternalApi private[akka] final class GroupedWithin[T](val n: Int, val d: FiniteDuration) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
-  require(n > 0, "n must be greater than 0")
+@InternalApi private[akka] final class GroupedWeightedWithin[T](val maxWeight: Long, costFn: T ⇒ Long, val d: FiniteDuration) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
+  require(maxWeight > 0, "maxWeight must be greater than 0")
   require(d > Duration.Zero)
 
   val in = Inlet[T]("in")
   val out = Outlet[immutable.Seq[T]]("out")
 
-  override def initialAttributes = DefaultAttributes.groupedWithin
+  override def initialAttributes = DefaultAttributes.groupedWeightedWithin
 
   val shape = FlowShape(in, out)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) with InHandler with OutHandler {
 
     private val buf: VectorBuilder[T] = new VectorBuilder
+    private var pending: T = null.asInstanceOf[T]
+    private var pendingWeight = 0L
     // True if:
     // - buf is nonEmpty
     //       AND
     // - timer fired OR group is full
     private var groupClosed = false
     private var groupEmitted = true
     private var finished = false
-    private var elements = 0
+    private var totalWeight = 0L
 
-    private val GroupedWithinTimer = "GroupedWithinTimer"
+    private val GroupedWeightedWithinTimer = "GroupedWeightedWithinTimer"
 
     override def preStart() = {
-      schedulePeriodically(GroupedWithinTimer, d)
+      schedulePeriodically(GroupedWeightedWithinTimer, d)
       pull(in)
     }
 
     private def nextElement(elem: T): Unit = {
       groupEmitted = false
-      buf += elem
-      elements += 1
-      if (elements == n) {
-        schedulePeriodically(GroupedWithinTimer, d)
+      val cost = costFn(elem)
+      if (totalWeight + cost <= maxWeight) {
+        buf += elem
+        totalWeight += cost
+        if (totalWeight == maxWeight) {
```

Comment:
Since weight can be `0`, we still need to grab new elements and add them while we still can.
Since downstream can request (and should get) when `weight == maxWeight`, we still have to check if it's available when timeout exceeded or `maxWeight` is reached, and emit, even if `pending == null`.